### PR TITLE
expose vector options

### PR DESCRIPTION
**Pull request by AviAvni**
fix https://github.com/FalkorDB/FalkorDB/issues/766

## Summary by CodeRabbit

- **New Features**
	- Introduced support for version 15 of the graph serialization protocol, including functions for encoding and decoding graph structures and entities.
	- Enhanced vector indexing with new configuration parameters for the HNSW algorithm, improving performance and flexibility.
	- Added options for indexed fields, allowing users to specify configurations for better data handling.
	- Improved graph indexing capabilities with new functions to create vector indexes based on embeddings.
	- Enhanced testing framework to validate the new indexing features and configurations.
	- Updated CodeQL action in the GitHub Actions workflow for enhanced stability and specific features.

- **Bug Fixes**
	- Improved memory management for graph context handling, reducing reallocations during graph data loading.

### CODE REVIEW

File: src/index/index.c [hunk 1]
```diff
@@ -140,6 +140,9 @@ static void _Index_ConstructStructure
 					field->vector_name);
 
 			RediSearch_VectorFieldSetDim(rsIdx, fieldID, field->options.dimension);
+			RediSearch_VectorFieldSetM(rsIdx, fieldID, field->options.M);
+			RediSearch_VectorFieldSetefConstruction(rsIdx, fieldID, field->options.efConstruction);
+			RediSearch_VectorFieldSetefRuntime(rsIdx, fieldID, field->options.efRuntime);
```
Comment (swilly22) [inline src/index/index.c]: would be nice to have a short comment explaining each parameter

File: src/index/index_field.c [hunk 2]
```diff
@@ -52,10 +52,13 @@ void IndexField_Init
...
-	field->options.nostem    = INDEX_FIELD_DEFAULT_NOSTEM;
-	field->options.phonetic  = rm_strdup(INDEX_FIELD_DEFAULT_PHONETIC);
-	field->options.dimension = 0;
+	field->options.weight         = INDEX_FIELD_DEFAULT_WEIGHT;
+	field->options.nostem         = INDEX_FIELD_DEFAULT_NOSTEM;
+	field->options.phonetic       = rm_strdup(INDEX_FIELD_DEFAULT_PHONETIC);
+	field->options.dimension      = 0;
+	field->options.M              = 16;
+	field->options.efConstruction = 200;
+	field->options.efRuntime      = 10;
```
Comment (swilly22) [inline src/index/index_field.c:61]: Do we know the purpose of each option?

File: src/index/index_field.c [hunk 3]
```diff
@@ -126,13 +129,19 @@ void IndexField_NewFullTextField
...
 (
-	IndexField *field,   // field to initialize
-	const char *name,    // field name
-	AttributeID id,      // field id
-	uint32_t dimension   // vector dimension
+	IndexField *field,      // field to initialize
+	const char *name,       // field name
+	AttributeID id,         // field id
+	uint32_t dimension,     // vector dimension
+	size_t M,		        // max outgoing edges
```
Comment (swilly22) [inline src/index/index_field.c:136]: comment alignment

File: src/index/index_field.h [hunk 4]
```diff
@@ -31,10 +31,13 @@ typedef struct {
...
 	IndexFieldType type;     // field type(s)
 	struct {
-		double weight;       // the importance of text
-		bool nostem;         // disable stemming of the text
-		char *phonetic;      // phonetic search of text
-		uint32_t dimension;  // vector dimension
+		double weight;         // the importance of text
+		bool nostem;           // disable stemming of the text
+		char *phonetic;        // phonetic search of text
+		uint32_t dimension;    // vector dimension
```
Comment (swilly22) [inline src/index/index_field.h:37]: would you consider grouping vector field options within an inner struct? e.g. HNSW params?

File: src/index/index_field.h [hunk 5]
```diff
@@ -80,10 +83,13 @@ void IndexField_NewFullTextField
...
 (
-	IndexField *field,   // field to initialize
-	const char *name,    // field name
-	AttributeID id,      // field id
-	uint32_t dimension   // vector dimension
+	IndexField *field,      // field to initialize
+	const char *name,       // field name
+	AttributeID id,         // field id
+	uint32_t dimension,     // vector dimension
+	size_t M,		        // max outgoing edges
```
Comment (swilly22) [inline src/index/index_field.h:90]: comment alignment

File: src/index/index_vector_create.c [hunk 6]
```diff
@@ -55,6 +61,33 @@ static bool _parseOptions
 		return false;
 	}
 
+	if(MAP_GET(options, "M", val)) {
+		if(SI_TYPE(val) != T_INT64) {
+			return false;
+		}
+		*M = val.longval;
+	} else {
+		*M = 16;
```
Comment (swilly22) [inline src/index/index_vector_create.c:70]: Consider using #define foreach default config value

File: src/index/index_field.c (hunk 2 above)
Comment (AviAvni) [reply src/index/index_field.c:61]: M - how much neighbor in the graph
efConstruction - how much vectors to compare in construction time
efRuntime - how much vectors to compare in query time

File: src/procedures/proc_list_indexes.c [hunk 7]
```diff
@@ -56,6 +58,12 @@ static void _process_yield
 			continue;
 		}
 
+		if(strcasecmp("options", yield[i]) == 0) {
```
Comment (swilly22) [inline src/procedures/proc_list_indexes.c:61]: I think 'options' is too generic of aname, as we're dealing with vector index options, I think the name should reflect it, consider renaming to 'HNSW_config'.

File: src/procedures/proc_list_indexes.c [hunk 8]
```diff
@@ -248,6 +256,45 @@ static bool _EmitIndex
...
 
+	//--------------------------------------------------------------------------
+	// index fields options
+	//--------------------------------------------------------------------------
+
+	if(ctx->yield_options) {
+		// {field_name, {field_options}
+		// e.g.
+		// {
+		//  'embedding': {'dimension': 100, 'M': 16, 'efConstruction': 200, 'efRuntime': 100}
```
Comment (swilly22) [inline src/procedures/proc_list_indexes.c:267]: why did you choose to call this section: 'embedding'?

File: src/serializers/encoder/v15/encode_schema.c [hunk 9]
```diff
@@ -63,10 +63,10 @@ static void _RdbSaveIndexField
...
 
 	// encode field dimension
-	SerializerIO_WriteUnsigned(rdb, f->options.dimension);
-	SerializerIO_WriteUnsigned(rdb, f->options.M);
-	SerializerIO_WriteUnsigned(rdb, f->options.efConstruction);
-	SerializerIO_WriteUnsigned(rdb, f->options.efRuntime);
+	SerializerIO_WriteUnsigned(rdb, f->hnsw_options.dimension);
+	SerializerIO_WriteUnsigned(rdb, f->hnsw_options.M);
+	SerializerIO_WriteUnsigned(rdb, f->hnsw_options.efConstruction);
+	SerializerIO_WriteUnsigned(rdb, f->hnsw_options.efRuntime);
```
Comment (swilly22) [inline src/serializers/encoder/v15/encode_schema.c:69]: consider using the getter functions

File: tests/flow/index_utils.py [hunk 10]
```diff
@@ -64,8 +64,8 @@ def create_node_range_index(graph, label, *properties, sync=False):
 def create_node_fulltext_index(graph, label, *properties, sync=False):
     return _create_typed_index(graph, "FULLTEXT", "NODE", label, *properties, sync=sync)
 
-def create_node_vector_index(graph, label, *properties, dim=0, similarity_function="euclidean", sync=False):
-    options = {'dimension': dim, 'similarityFunction': similarity_function}
+def create_node_vector_index(graph, label, *properties, dim=0, similarity_function="euclidean", m=16, efConstruction=200, efRuntime=10, sync=False):
```
Comment (swilly22) [inline tests/flow/index_utils.py:67]: how did you came up with these default values?

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/procedures/proc_list_indexes.c (hunk 8 above)
Comment (AviAvni) [reply src/procedures/proc_list_indexes.c:267]: embedding is the field name

File: src/procedures/proc_list_indexes.c (hunk 7 above)
Comment (AviAvni) [reply src/procedures/proc_list_indexes.c:61]: this is for any options if we will introduce more options for the other index type you suggest to add more variables to the result? for example range_options and fulltext_options?

File: src/procedures/proc_list_indexes.c (hunk 8 above)
Comment (swilly22) [reply src/procedures/proc_list_indexes.c:267]: I under this is the field name but the value describes a vector index config, why not call this vector_config or vector_parameters?

File: src/procedures/proc_list_indexes.c (hunk 7 above)
Comment (swilly22) [reply src/procedures/proc_list_indexes.c:61]: So this is general options for all index configuration ?

**Review by swilly22 — APPROVED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**